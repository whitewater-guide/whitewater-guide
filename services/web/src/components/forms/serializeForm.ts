import { toMarkdown } from '@whitewater-guide/md-editor';
import mapValues from 'lodash/mapValues';

/**
 * Convert form data into graphql mutation input
 * @param {string[]} markdownFields Fields that must be converted from draft.js internal
 * representation to markdown strings
 * @param {string[]} refs one-to-one connections, should be reduced to { id } objects
 * @param {string[]} connections one-to-many connections, should be reduced to [ { id }, { id } ] arrays
 * @returns {(input?: (object | null)) => (any | {[p: string]: any})} serializer function
 */
export const serializeForm = (
  markdownFields: string[] = [],
  refs: string[] = [],
  connections: string[] = [],
) => (input?: object | null) => {
  if (!input) {
    return null;
  }
  return mapValues(input, (value: any, key: string) => {
    if (markdownFields.includes(key)) {
      return value ? toMarkdown(value) : null;
    } else if (refs.includes(key)) {
      return value ? { id: value.id } : null;
    } else if (connections.includes(key)) {
      return value.map((item: any) => ({ id: item.id }));
    }
    return value;
  }) as { [key: string]: any };
};
