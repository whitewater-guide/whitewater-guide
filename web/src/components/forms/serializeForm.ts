import { stateToMarkdown } from 'draft-js-export-markdown';
import { mapValues } from 'lodash';

export const serializeForm = (markdownFields: string[]) => (input?: object | null) => {
  if (!input) {
    return null;
  }
  return mapValues(input, (value: any, key: string) => {
    if (markdownFields.includes(key)) {
      return value ? stateToMarkdown(value.getCurrentContent()).trim() : null;
    }
    return value;
  });
};
