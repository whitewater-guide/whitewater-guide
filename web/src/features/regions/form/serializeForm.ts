import draftToMarkdown from 'draft-js-export-markdown';
import { RegionInput } from '../../../ww-commons';
import { RegionFormInput } from './types';

export default function serializeForm(formData?: RegionFormInput): RegionInput | null {
  if (!formData) {
    return null;
  }
  const { description } = formData;
  const stringDescription = description ?
    draftToMarkdown(description.getCurrentContent()) : null;
  return {
    ...formData,
    description: stringDescription,
  };
}
