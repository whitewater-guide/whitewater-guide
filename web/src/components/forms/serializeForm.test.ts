import { ContentState, EditorState } from 'draft-js';
import { serializeForm } from './serializeForm';

const input = {
  id: null,
  name: 'aa',
  pois: [],
  description: null,
  seasonNumeric: [],
  season: null,
  bounds: null,
  hidden: false,
};

const inputWithDraft = {
  ...input,
  description: EditorState.createWithContent(ContentState.createFromText('foo')),
};

const serializer = serializeForm(['description']);

it('should handle undefined input', () => {
  expect(serializer(undefined)).toBeNull();
});

it('should serialize empty markdown field', () => {
  expect(serializer(input)!.description).toBeNull();
});

it('should transform draft.js into markdown string', () => {
  expect(serializer(inputWithDraft)!.description).toBe('foo');
});

it('should match snapshot', () => {
  expect(serializer(inputWithDraft)).toMatchSnapshot();
});
