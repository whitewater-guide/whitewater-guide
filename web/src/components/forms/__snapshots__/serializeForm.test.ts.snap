// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
Object {
  "bounds": null,
  "description": "foo",
  "hidden": false,
  "id": null,
  "name": "aa",
  "pois": Array [],
  "season": null,
  "seasonNumeric": Array [],
}
`;
